cmd/thanos/receive.go

```diff
@@ -321,23 +359,78 @@ func runReceive(
 
 		s := shipper.New(logger, reg, dataDir, bkt, func() labels.Labels { return lset }, metadata.ReceiveSource)
 
-		ctx, cancel := context.WithCancel(context.Background())
-		g.Add(func() error {
-			defer runutil.CloseWithLogOnErr(logger, bkt, "bucket client")
+		// Before starting, ensure any old blocks are uploaded.
+		if uploaded, err := s.Sync(context.Background()); err != nil {
+			level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+		}
+		// Before quitting, ensure all blocks are uploaded.
+		defer func() {
+			if uploaded, err := s.Sync(context.Background()); err != nil {
+				level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+			}
+		}()
 
-			return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
-				if uploaded, err := s.Sync(ctx); err != nil {
-					level.Warn(logger).Log("err", err, "uploaded", uploaded)
-				}
+		{
+			// Run the uploader in a loop.
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				return runutil.Repeat(30*time.Second, ctx.Done(), func() error {
+					if uploaded, err := s.Sync(ctx); err != nil {
+						level.Warn(logger).Log("err", err, "uploaded", uploaded)
+					}
 
-				return nil
+					return nil
+				})
+			}, func(error) {
+				cancel()
 			})
-		}, func(error) {
-			cancel()
-		})
+		}
+
+		{
+			// Upload on demand.
+			ctx, cancel := context.WithCancel(context.Background())
+			g.Add(func() error {
+				for {
+					select {
+					case <-ctx.Done():
+						return nil
+					case <-uploadC:
+						if uploaded, err := s.Sync(ctx); err != nil {
+							level.Warn(logger).Log("err", err, "failed to upload", uploaded)
+						}
+						cancel()
+						sendOrCancel(ctx.Done(), uploadDone)
+					}
+				}
+			}, func(error) {
+				cancel()
+				close(uploadDone)
+			})
+		}
 	}
 
+	// Before quitting, ensure the WAL is flushed.
+	defer func() {
+		if err := db.Open(); err != nil {
+			level.Warn(logger).Log("err", err, "msg", "failed to open storage")
+			return
+		}
+		if err := db.Flush(); err != nil {
+			level.Warn(logger).Log("err", err, "msg", "failed to flush storage")
+			return
+		}
+	}()
+
 	level.Info(logger).Log("msg", "starting receiver")
 
 	return nil
 }
+
+// sendOrCancel allows us to signal a chan and still
+// quit early if the receiver has already quit.
+func sendOrCancel(cancel <-chan struct{}, ch chan<- struct{}) {
+	select {
+	case <-cancel:
```

Comment:
the idea here was to let the send be cancel-able, not simply non-blocking.

Reply:
got it thanks.

Reply:
The cancelable contexts by themselves are not enough, I think. Take the case of the run group managing the DB. Whenever the DB gets updated, it needs to be flushed and then it needs to signal to uploader that it should upload the blocks. Here the uploading and flushing is synchronous but not preemptible:
```go
for {
    select {
    case <- cancel:
        return nil
    case <- dbUpdate:
        db.Flush()
        uploadC <- struct{}{}
        <-uploadDone
    }
}
```
The uploading is blocking, this means that if we need want to cancel the run group for the db manager, it will have to wait for uploading to finish. Also, if the uploader has already exited, then no one will every receive on `uploadC` and the flushing run group will be blocked forever. However, thinking about this again, I think we can get away with simply using buffered chans. This will ensure we can synchronize the run groups and still not block when we need to cancel. I'll post that change, which should simplify things a bit (: